### ISSUE TITLE
`extract_variable` should be able to heuristically extract multiple occurences of the same expression

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#10547]
As mentioned in https://github.com/rust-analyzer/rust-analyzer/issues/7692#issuecomment-888356714_ it might be nice to have `extract_variable` lift out all occurences of the target expression in the current function and replace them with the new variable name. 

Open question is what theres heuristics are exactly. For example this should keep in mind the places of these expressions to decide whether it would make sense to extract more than just the target, like expressions in function arguments where those parameters have the same name etc.

**Comment by jplatte** [comment]
For the user case I had when writing that comment, "extract to variable" heuristics wouldn't actually help since the occurences of the string were across multiple functions. Maybe a separate "extract to constant" assist would make sense afterall.

**Comment by ChayimFriedman2** [comment]
I don't think this pulls its weight, unless we limit it to the very simple case of extracting a literal, and even then maybe not.

**Comment by iago-lito** [comment]
I guess it would make better sense once (if ever) we get clear `no-side-effect` expression markers? I *think* it is under discussion at the lang level rn but not landing anytime soon.

**Comment by jplatte** [comment]
Extracting a literal seems like the most common case, so I'd still be happy to have this with that limitation.
